File: src/test/java/com/soapdogg/lispInterpreter/evaluator/CondFunctionEvaluatorTest.java
package com.soapdogg.lispInterpreter.evaluator;

import com.soapdogg.lispInterpreter.comparator.NodeValueComparator;
import com.soapdogg.lispInterpreter.constants.FunctionNameConstants;
import com.soapdogg.lispInterpreter.datamodels.UserDefinedFunction;
import com.soapdogg.lispInterpreter.exceptions.NotAListException;
import com.soapdogg.lispInterpreter.datamodels.AtomNode;
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode;
import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.valueretriver.ListValueRetriever;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.Collections;
import java.util.List;
import java.util.Map;

class CondFunctionEvaluatorTest {

    private Node params;
    private List<UserDefinedFunction> userDefinedFunctions;
    private Map<String, Node> variableNameToValueMap;

    private ListValueRetriever listValueRetriever;
    private NodeEvaluator nodeEvaluator;
    private NodeValueComparator nodeValueComparator;

    private CondFunctionEvaluator condFunctionEvaluator;

    @BeforeEach
    void setup() {
        userDefinedFunctions = Collections.emptyList();
        variableNameToValueMap = Collections.emptyMap();

        listValueRetriever = Mockito.mock(ListValueRetriever.class);
        nodeEvaluator = Mockito.mock(NodeEvaluator.class);
        nodeValueComparator = Mockito.mock(NodeValueComparator.class);

        condFunctionEvaluator = new CondFunctionEvaluator(
            listValueRetriever,
            nodeEvaluator,
            nodeValueComparator
        );
    }

    @Test
    void inputIsAtomNodeTest() {
        params = Mockito.mock(AtomNode.class);

        Assertions.assertThrows(
            NotAListException.class,
            () -> condFunctionEvaluator.evaluateCondFunction(
                params,
                userDefinedFunctions,
                variableNameToValueMap
            )
        );

        Mockito.verifyNoInteractions(listValueRetriever);
        Mockito.verifyNoInteractions(nodeEvaluator);
        Mockito.verifyNoInteractions(nodeValueComparator);
    }

    @Test
    void inputIsListAndBooleanResultIsNotNilTest() throws Exception {
        params = Mockito.mock(ExpressionNode.class);

        Node address = Mockito.mock(Node.class);
        Mockito.when(((ExpressionNode)params).getAddress()).thenReturn(address);

        ExpressionNode expressionNodeAddress = Mockito.mock(ExpressionNode.class);
        Mockito.when(
            listValueRetriever.retrieveListValue(
                address,
                FunctionNameConstants.COND,
                variableNameToValueMap
            )
        ).thenReturn(expressionNodeAddress);

        Node expressionNodeAddressAddress = Mockito.mock(Node.class);
        Mockito.when(expressionNodeAddress.getAddress()).thenReturn(expressionNodeAddressAddress);

        Node booleanResult = Mockito.mock(AtomNode.class);
        Mockito.when(
            nodeEvaluator.evaluate(
                expressionNodeAddressAddress,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(booleanResult);

        Mockito.when(
            nodeValueComparator.equalsNil(
                ((AtomNode)booleanResult).getValue()
            )
        ).thenReturn(false);

        Node expressionNodeAddressData = Mockito.mock(Node.class);
        Mockito.when(expressionNodeAddress.getData()).thenReturn(expressionNodeAddressData);

        Node expected = Mockito.mock(Node.class);
        Mockito.when(
            nodeEvaluator.evaluate(
                expressionNodeAddressData,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(expected);

        Node actual = condFunctionEvaluator.evaluateCondFunction(
            params,
            userDefinedFunctions,
            variableNameToValueMap
        );

        Assertions.assertEquals(expected, actual);
    }

    @Test
    void inputIsListAndBooleanResultIsNotAtomNodeTest() throws Exception {
        params = Mockito.mock(ExpressionNode.class);

        Node address = Mockito.mock(Node.class);
        Mockito.when(((ExpressionNode)params).getAddress()).thenReturn(address);

        ExpressionNode expressionNodeAddress = Mockito.mock(ExpressionNode.class);
        Mockito.when(
            listValueRetriever.retrieveListValue(
                address,
                FunctionNameConstants.COND,
                variableNameToValueMap
            )
        ).thenReturn(expressionNodeAddress);

        Node expressionNodeAddressAddress = Mockito.mock(Node.class);
        Mockito.when(expressionNodeAddress.getAddress()).thenReturn(expressionNodeAddressAddress);

        Node booleanResult = Mockito.mock(ExpressionNode.class);
        Mockito.when(
            nodeEvaluator.evaluate(
                expressionNodeAddressAddress,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(booleanResult);

        Node data = Mockito.mock(AtomNode.class);
        Mockito.when(((ExpressionNode)params).getData()).thenReturn(data);

        Assertions.assertThrows(
            NotAListException.class,
            () -> condFunctionEvaluator.evaluateCondFunction(
                params,
                userDefinedFunctions,
                variableNameToValueMap
            )
        );
    }

    @Test
    void inputIsListAndBooleanResultIsNilTest() throws Exception {
        params = Mockito.mock(ExpressionNode.class);

        Node address = Mockito.mock(Node.class);
        Mockito.when(((ExpressionNode)params).getAddress()).thenReturn(address);

        ExpressionNode expressionNodeAddress = Mockito.mock(ExpressionNode.class);
        Mockito.when(
            listValueRetriever.retrieveListValue(
                address,
                FunctionNameConstants.COND,
                variableNameToValueMap
            )
        ).thenReturn(expressionNodeAddress);

        Node expressionNodeAddressAddress = Mockito.mock(Node.class);
        Mockito.when(expressionNodeAddress.getAddress()).thenReturn(expressionNodeAddressAddress);

        Node booleanResult = Mockito.mock(AtomNode.class);
        Mockito.when(
            nodeEvaluator.evaluate(
                expressionNodeAddressAddress,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(booleanResult);

        Mockito.when(
            nodeValueComparator.equalsNil(
                ((AtomNode)booleanResult).getValue()
            )
        ).thenReturn(true);

        Node data = Mockito.mock(AtomNode.class);
        Mockito.when(((ExpressionNode)params).getData()).thenReturn(data);

        Assertions.assertThrows(
            NotAListException.class,
            () -> condFunctionEvaluator.evaluateCondFunction(
                params,
                userDefinedFunctions,
                variableNameToValueMap
            )
        );
    }
}

File: src/test/java/com/soapdogg/lispInterpreter/evaluator/CondFunctionEvaluatorTest.kt
package com.soapdogg.lispInterpreter.evaluator

import com.soapdogg.lispInterpreter.comparator.NodeValueComparator
import com.soapdogg.lispInterpreter.constants.FunctionNameConstants
import com.soapdogg.lispInterpreter.datamodels.AtomNode
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode
import com.soapdogg.lispInterpreter.datamodels.Node
import com.soapdogg.lispInterpreter.datamodels.UserDefinedFunction
import com.soapdogg.lispInterpreter.exceptions.NotAListException
import com.soapdogg.lispInterpreter.valueretriver.ListValueRetriever
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class CondFunctionEvaluatorTest {
    private val userDefinedFunctions: List<UserDefinedFunction> = emptyList()
    private val variableNameToValueMap: Map<String, Node> = emptyMap()
    private val listValueRetriever = Mockito.mock(ListValueRetriever::class.java)
    private val nodeEvaluator = Mockito.mock(NodeEvaluator::class.java)
    private val nodeValueComparator = Mockito.mock(NodeValueComparator::class.java)
    private val condFunctionEvaluator = CondFunctionEvaluator(
        listValueRetriever,
        nodeEvaluator,
        nodeValueComparator
    )

    @Test
    fun inputIsAtomNodeTest() {
        val params: Node = Mockito.mock(AtomNode::class.java)
        Assertions.assertThrows(
            NotAListException::class.java
        ) {
            condFunctionEvaluator.evaluateCondFunction(
                params,
                userDefinedFunctions,
                variableNameToValueMap
            )
        }
        Mockito.verifyNoInteractions(listValueRetriever)
        Mockito.verifyNoInteractions(nodeEvaluator)
        Mockito.verifyNoInteractions(nodeValueComparator)
    }

    @Test
    fun inputIsListAndBooleanResultIsNotNilTest() {
        val params: Node = Mockito.mock(ExpressionNode::class.java)
        val address = Mockito.mock(Node::class.java)
        Mockito.`when`((params as ExpressionNode).address).thenReturn(address)
        val expressionNodeAddress = Mockito.mock(ExpressionNode::class.java)
        Mockito.`when`(
            listValueRetriever.retrieveListValue(
                address,
                FunctionNameConstants.COND,
                variableNameToValueMap
            )
        ).thenReturn(expressionNodeAddress)
        val expressionNodeAddressAddress = Mockito.mock(Node::class.java)
        Mockito.`when`(expressionNodeAddress.address).thenReturn(expressionNodeAddressAddress)
        val booleanResult: Node = Mockito.mock(AtomNode::class.java)
        Mockito.`when`(
            nodeEvaluator.evaluate(
                expressionNodeAddressAddress,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(booleanResult)
        Mockito.`when`(
            nodeValueComparator.equalsNil(
                (booleanResult as AtomNode).value
            )
        ).thenReturn(false)
        val expressionNodeAddressData = Mockito.mock(Node::class.java)
        Mockito.`when`(expressionNodeAddress.data).thenReturn(expressionNodeAddressData)
        val expected = Mockito.mock(Node::class.java)
        Mockito.`when`(
            nodeEvaluator.evaluate(
                expressionNodeAddressData,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(expected)
        val actual = condFunctionEvaluator.evaluateCondFunction(
            params,
            userDefinedFunctions,
            variableNameToValueMap
        )
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun inputIsListAndBooleanResultIsNotAtomNodeTest() {
        val params: Node = Mockito.mock(ExpressionNode::class.java)
        val address = Mockito.mock(Node::class.java)
        Mockito.`when`((params as ExpressionNode).address).thenReturn(address)
        val expressionNodeAddress = Mockito.mock(ExpressionNode::class.java)
        Mockito.`when`(
            listValueRetriever.retrieveListValue(
                address,
                FunctionNameConstants.COND,
                variableNameToValueMap
            )
        ).thenReturn(expressionNodeAddress)
        val expressionNodeAddressAddress = Mockito.mock(Node::class.java)
        Mockito.`when`(expressionNodeAddress.address).thenReturn(expressionNodeAddressAddress)
        val booleanResult: Node = Mockito.mock(ExpressionNode::class.java)
        Mockito.`when`(
            nodeEvaluator.evaluate(
                expressionNodeAddressAddress,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(booleanResult)
        val data: Node = Mockito.mock(AtomNode::class.java)
        Mockito.`when`(params.data).thenReturn(data)
        Assertions.assertThrows(
            NotAListException::class.java
        ) {
            condFunctionEvaluator.evaluateCondFunction(
                params,
                userDefinedFunctions,
                variableNameToValueMap
            )
        }
    }

    @Test
    fun inputIsListAndBooleanResultIsNilTest() {
        val params: Node = Mockito.mock(ExpressionNode::class.java)
        val address = Mockito.mock(Node::class.java)
        Mockito.`when`((params as ExpressionNode).address).thenReturn(address)
        val expressionNodeAddress = Mockito.mock(ExpressionNode::class.java)
        Mockito.`when`(
            listValueRetriever.retrieveListValue(
                address,
                FunctionNameConstants.COND,
                variableNameToValueMap
            )
        ).thenReturn(expressionNodeAddress)
        val expressionNodeAddressAddress = Mockito.mock(Node::class.java)
        Mockito.`when`(expressionNodeAddress.address).thenReturn(expressionNodeAddressAddress)
        val booleanResult: Node = Mockito.mock(AtomNode::class.java)
        Mockito.`when`(
            nodeEvaluator.evaluate(
                expressionNodeAddressAddress,
                userDefinedFunctions,
                variableNameToValueMap,
                true
            )
        ).thenReturn(booleanResult)
        Mockito.`when`(
            nodeValueComparator.equalsNil(
                (booleanResult as AtomNode).value
            )
        ).thenReturn(true)
        val data: Node = Mockito.mock(AtomNode::class.java)
        Mockito.`when`(params.data).thenReturn(data)
        Assertions.assertThrows(
            NotAListException::class.java
        ) {
            condFunctionEvaluator.evaluateCondFunction(
                params,
                userDefinedFunctions,
                variableNameToValueMap
            )
        }
    }
}